open-craft/openedx-completion-aggregator
# [MCKIN-5899] Create model to track aggregate completion (was edx:edx-platform#16715)

## Creates an Aggregate completion model for the completion API.

This is moved from https://github.com/edx/edx-platform/pull/16715, because after discussion with @iloveagent57 and @schenedx over at edX, we are moving the aggregation code to an external repo from edx-platform.

@macornwell - Please look over this to see that it still matches your intentions from the original PR.

### Changes from the original PR:

* Moved to external repo
* Renamed CompletionAggregation to Aggregator, because the app name already tells that it's completion-related.
* Removed Waffle flag, because edx can disable it simply by not installing it in edx platform.
* Made id field a regular id, because BigIntAuto was dependent on edx platform and not neccesary for this model.  (There will be far fewer of these than raw BlockCompletions).
* Vendored and simplified CourseKeyField and UsageKeyField, because they are in edx-platform
* Added tests around custom fields
* Removed tests around waffle flags
* Introduced python3 compatibility decorator on model.

### Original PR description

The AggregateCompletion model inherits from TimeStampedModel and has the following fields:
```
id = BigAutoField(primary_key=True)  # pylint: disable=invalid-name
user = models.ForeignKey(User)
course_key = CourseKeyField(max_length=255)
aggregation_name = models.Char
block_key = UsageKeyField(max_length=255)
earned = models.FloatField(validators=[validate_positive_float])
possible = models.FloatField(validators=[validate_positive_float])
percent = models.FloatField(validators=[validate_percent])
```


**JIRA tickets**: MCKIN-5899

**Sandbox URL**:
- LMS: https://pr16715.sandbox.opencraft.hosting
- Studio: https://studio-pr16715.sandbox.opencraft.hosting

**Merge deadline**: None

**Testing instructions**:
1. SSH into a devstack shell.
2. Install openedx-completion-aggregator:
```
pip install git+https://github.com/open-craft/openedx-completion-aggregator.git@cornwell/aggregator-models#egg=openedx-completion-aggregator==0.1
```
3. Add completion_aggregator to lms/envs/common.py:INSTALLED_APPS, and ensure the database has been migrated.
```
python manage.py lms migrate --settings=devstack
```
3. Start a python shell.
edxapp@vagrant:~/edx-platform$ python manage.py lms shell --settings=devstack
```python manage.py lms shell --settings=devstack```

4. Run the following python commands to create a new Aggregator object.
```
from completion_aggregator.models import Aggregator;
from opaque_keys.edx.keys import UsageKey;
from django.contrib.auth import get_user_model
user = get_user_model().objects.get(username='audit')
block_key_obj = UsageKey.from_string(u'block-v1:edx+test+run+type@video+block@doggos');
Aggregator.objects.submit_completion(user, block_key_obj.course_key, block_key_obj, "course", 0.5, 1);
```
5. Result should be the following:
```
(<Aggregator: Aggregator: audit, course-v1:edx+test+run, block-v1:edx+test+run+type@video+block@doggos: 0.5>, True)
```

**Reviewers**
